File: frontend/src/utils/coordinate.ts
const originThreshold = 0.1;

export function isValidCoordinate(longitude: number, latitude: number) {
  if (!Number.isFinite(longitude) || !Number.isFinite(latitude)) {
    return false;
  }
  if (longitude < -180 || longitude > 180 || latitude < -90 || latitude > 90) {
    return false;
  }
  const nearOrigin = Math.abs(longitude) < originThreshold && Math.abs(latitude) < originThreshold;
  return !nearOrigin && longitude !== 0 && latitude !== 0;
}

File: internal/coordinate/coordinate.go
// Package coordinate provides shared geographic coordinate validation.
package coordinate

import "math"

const originThreshold = 0.1

// IsValid reports whether longitude and latitude contain usable WGS84 coordinates.
func IsValid(longitude, latitude float64) bool {
	if math.IsNaN(longitude) || math.IsInf(longitude, 0) ||
		math.IsNaN(latitude) || math.IsInf(latitude, 0) {
		return false
	}
	if longitude < -180 || longitude > 180 || latitude < -90 || latitude > 90 {
		return false
	}

	nearOrigin := math.Abs(longitude) < originThreshold && math.Abs(latitude) < originThreshold
	return !nearOrigin && longitude != 0 && latitude != 0
}

// IsValidCoord reports whether non-nil longitude and latitude pointers contain usable WGS84 coordinates.
func IsValidCoord(longitude, latitude *float64) bool {
	return longitude != nil && latitude != nil && IsValid(*longitude, *latitude)
}
